substrate/frame/election-provider-support/src/onchain.rs

```diff
@@ -80,78 +95,64 @@ pub trait Config {
-		// early exit
-		return Err(Error::TooManyWinners)
-	}
+impl<T: Config> OnChainExecution<T> {
+	fn elect_with(
+		bounds: ElectionBounds,
+		page: PageIndex,
+	) -> Result<BoundedSupportsOf<Self>, Error> {
+		let (voters, targets) = T::DataProvider::electing_voters(bounds.voters, page)
+			.and_then(|voters| {
+				Ok((voters, T::DataProvider::electable_targets(bounds.targets, page)?))
+			})
+			.map_err(Error::DataProvider)?;
+
+		let desired_targets = T::DataProvider::desired_targets().map_err(Error::DataProvider)?;
+
+		if desired_targets > T::MaxWinnersPerPage::get() {
+			// early exit
+			return Err(Error::TooManyWinners)
+		}
 
-	let voters_len = voters.len() as u32;
-	let targets_len = targets.len() as u32;
+		let voters_len = voters.len() as u32;
+		let targets_len = targets.len() as u32;
 
-	let stake_map: BTreeMap<_, _> = voters
-		.iter()
-		.map(|(validator, vote_weight, _)| (validator.clone(), *vote_weight))
-		.collect();
+		let stake_map: BTreeMap<_, _> = voters
+			.iter()
+			.map(|(validator, vote_weight, _)| (validator.clone(), *vote_weight))
+			.collect();
 
-	let stake_of = |w: &<T::System as frame_system::Config>::AccountId| -> VoteWeight {
-		stake_map.get(w).cloned().unwrap_or_default()
-	};
+		let stake_of = |w: &<T::System as frame_system::Config>::AccountId| -> VoteWeight {
+			stake_map.get(w).cloned().unwrap_or_default()
+		};
 
-	let ElectionResult { winners: _, assignments } =
-		T::Solver::solve(desired_targets as usize, targets, voters).map_err(Error::from)?;
+		let ElectionResult { winners: _, assignments } =
+			T::Solver::solve(desired_targets as usize, targets, voters).map_err(Error::from)?;
 
-	let staked = assignment_ratio_to_staked_normalized(assignments, &stake_of)?;
+		let staked = assignment_ratio_to_staked_normalized(assignments, &stake_of)?;
 
-	let weight = T::Solver::weight::<T::WeightInfo>(
-		voters_len,
-		targets_len,
-		<T::DataProvider as ElectionDataProvider>::MaxVotesPerVoter::get(),
-	);
-	frame_system::Pallet::<T::System>::register_extra_weight_unchecked(
-		weight,
-		DispatchClass::Mandatory,
-	);
+		let weight = T::Solver::weight::<T::WeightInfo>(
+			voters_len,
+			targets_len,
+			<T::DataProvider as ElectionDataProvider>::MaxVotesPerVoter::get(),
+		);
+		frame_system::Pallet::<T::System>::register_extra_weight_unchecked(
+			weight,
+			DispatchClass::Mandatory,
+		);
 
-	// defensive: Since npos solver returns a result always bounded by `desired_targets`, this is
-	// never expected to happen as long as npos solver does what is expected for it to do.
-	let supports: OnChainBoundedSupportsOf<T> =
-		to_supports(&staked).try_into().map_err(|_| Error::TooManyWinners)?;
-
-	Ok(supports)
-}
+		// defensive: Since npos solver returns a result always bounded by `desired_targets`, this
+		// is never expected to happen as long as npos solver does what is expected for it to do.
+		let supports: BoundedSupportsOf<Self> = to_supports(&staked)
+			.try_into_bounded_supports()
```

Comment:
okay, so this conversion checks the bounds, and bails if any are not met. So the onchain election has no trimming capabiliites here. 

Here this is fine. But long term the MB-Miner should ideally be able to trim to meet these bounds, please make a note of this if it is not the case already.

Reply:
Keeping track of this here https://github.com/paritytech/polkadot-sdk/issues/6237, I'll ensure that's the case for all the (offchain) miners.